### ISSUE TITLE
How to convert `ScalableTag<double>` to `Rebind<size_t,...>`

### ISSUE THREAD

**Issue by fbarbari** [google/highway#1662]
As the title suggests, I am trying to vectorize a function where the result of a computation on `double`s is cast to `size_t` and then used as an index.
Here is the original snippet:
```
const point3D min = grid.min;
const double grid_factor = grid.grid_factor;
...

for(size_t i{0}; i < n; i++) {
    const double x = atoms.x[i];
    const double y = atoms.y[i];
    const double z = atoms.z[i];

    const size_t x_index = static_cast<size_t>((x - min.x) * grid_factor);
    const size_t y_index = static_cast<size_t>((y - min.y) * grid_factor);
    const size_t z_index = static_cast<size_t>((z - min.z) * grid_factor);
...
```

So, I tried to vectorize it this way:
```
namespace hn = hwy::HWY_NAMESPACE;
const hn::ScalableTag<double> coordinate_vector_tag;
using coordinate_vector = hn::Vec<decltype(coordinate_vector_tag)>;
const hn::Rebind<size_t, decltype(coordinate_vector_tag)> index_vector_tag;
using index_vector = hn::Vec<decltype(index_vector_tag)>;

const size_t num_lanes = static_cast<size_t>(hn::Lanes(coordinate_vector_tag));

const coordinate_vector min_x = hn::Set(coordinate_vector_tag, grid.min.x);
const coordinate_vector min_y = hn::Set(coordinate_vector_tag, grid.min.y);
const coordinate_vector min_z = hn::Set(coordinate_vector_tag, grid.min.z);
const coordinate_vector v_grid_factor = hn::Set(coordinate_vector_tag, grid.grid_factor);
...
for(size_t i{0}; i < n; i++) {
    const coordinate_vector x_i = hn::LoadU(coordinate_vector_tag, atom.x + i);
    const coordinate_vector y_i = hn::LoadU(coordinate_vector_tag, atom.y + i);
    const coordinate_vector z_i = hn::LoadU(coordinate_vector_tag, atom.z + i);

    const coordinate_vector x_i_index_double = hn::Mul(hn::Sub(x_i, min_x), v_grid_factor);
    const coordinate_vector y_i_index_double = hn::Mul(hn::Sub(y_i, min_y), v_grid_factor);
    const coordinate_vector z_i_index_double = hn::Mul(hn::Sub(z_i, min_z), v_grid_factor);
...
```

And then I need to convert from `{x,y,z}_i_index_double` to `{x,y,z}_i_index` but I can't find a way.
I tried using an `int64_t` with `DemoteTo(index_vector_tag, ConvertTo(Rebind<int64_t, decltype(coordinate_vector_tag)>(), x_i_index_double))` and a `float` with `ConvertTo(index_vector_tag, DemoteTo(Rebind<float, decltype(coordinate_vector_tag)>(), x_i_index_double))`, but both don't compile.

**Comment by johnplatts** [comment]
`const hn::Rebind<size_t, decltype(coordinate_vector_tag)> index_vector_tag;` needs to be replaced with the following:
```using TIdx = hwy::UnsignedFromSize<sizeof(size_t)>;
const hn::Rebind<TIdx, decltype(coordinate_vector_tag)> index_vector_tag;
```

Note that there are some architectures such as x86_32 and ARMv7 where size_t is a 32-bit unsigned integer and there are some other architectures such as x86_64, powerpc64, AArch64, and riscv64 where size_t is a 64-bit unsigned integer.

**Comment by johnplatts** [comment]
Here are some helper function that can be used to convert source values to the correct type:
```
template <class D, class V, class V2 = hn::Vec<hn::Rebind<hn::TFromV<V>, D>>,
          hwy::EnableIf<(sizeof(hn::TFromD<D>) < sizeof(hn::TFromV<V>) &&
                         HWY_MAX_LANES_D(hn::DFromV<V>) ==
                             HWY_MAX_LANES_D(hn::DFromV<V2>))>* = nullptr>
static HWY_INLINE hn::Vec<D> ConvertValues(D d, V v) {
  return hn::DemoteTo(d, v);
}

template <
    class D, class V, class V2 = hn::Vec<hn::Rebind<hn::TFromV<V>, D>>,
    hwy::EnableIf<(sizeof(hn::TFromD<D>) == sizeof(hn::TFromV<V>) &&
                   HWY_MAX_LANES_D(hn::DFromV<V>) ==
                       HWY_MAX_LANES_D(hn::DFromV<V2>)&&(
                           hwy::IsFloat<hn::TFromD<D>>() ||
                           hwy::IsSpecialFloat<hn::TFromD<D>>()) !=
                       (hwy::IsFloat<hn::TFromV<V>>() ||
                        hwy::IsSpecialFloat<hn::TFromV<V>>()))>* = nullptr>
static HWY_INLINE hn::Vec<D> ConvertValues(D d, V v) {
  return hn::ConvertTo(d, v);
}

template <class D, class V, class V2 = hn::Vec<hn::Rebind<hn::TFromV<V>, D>>,
          hwy::EnableIf<(sizeof(hn::TFromD<D>) == sizeof(hn::TFromV<V>) &&
                         HWY_MAX_LANES_D(hn::DFromV<V>) ==
                             HWY_MAX_LANES_D(hn::DFromV<V2>)&&(
                                 hwy::IsSame<hn::TFromD<D>, hn::TFromV<V>>() ||
                                 (!hwy::IsFloat<hn::TFromD<D>>() &&
                                  !hwy::IsSpecialFloat<hn::TFromD<D>>() &&
                                  !hwy::IsFloat<hn::TFromV<V>>() &&
                                  !hwy::IsSpecialFloat<hn::TFromV<V>>())))>* =
              nullptr>
static HWY_INLINE hn::Vec<D> ConvertValues(D d, V v) {
  return hn::BitCast(d, v);
}

template <class D, class V, class V2 = hn::Vec<hn::Rebind<hn::TFromV<V>, D>>,
          hwy::EnableIf<(sizeof(hn::TFromD<D>) > sizeof(hn::TFromV<V>) &&
                         HWY_MAX_LANES_D(hn::DFromV<V>) ==
                             HWY_MAX_LANES_D(hn::DFromV<V2>))>* = nullptr>
static HWY_INLINE hn::Vec<D> ConvertValues(D d, V v) {
  return hn::PromoteTo(d, v);
}
```

**Comment by fbarbari** [comment]
Thanks @johnplatts for the quick response.
Regarding your first answer, what should `TIdx` be? I understand the issue with `size_t` but, unfortunately, I cannot make such a radical change in the codebase I am working on right now. Moreover, this application is intended to be used only on x86_64 and AArch64, for now.
Regarding the second answer, I tried copy-pasting your code but it doesn't work. I tried setting `TIdx` to `size_t`, `long unsigned int` and `uint64_t` but none of them works.
Here is the code: [https://godbolt.org/z/se8TjovqK](https://godbolt.org/z/se8TjovqK)

**Comment by johnplatts** [comment]
> Thanks @johnplatts for the quick response. Regarding your first answer, what should `TIdx` be? I understand the issue with `size_t` but, unfortunately, I cannot make such a radical change in the codebase I am working on right now. Moreover, this application is intended to be used only on x86_64 and AArch64, for now. Regarding the second answer, I tried copy-pasting your code but it doesn't work. I tried setting `TIdx` to `size_t`, `long unsigned int` and `uint64_t` but none of them works. Here is the code: https://godbolt.org/z/se8TjovqK

Here is a link to the updated version of your snippet, which will correctly compile:
https://godbolt.org/z/dvWsjzsdr

I have also added some additional floating-point to integer and integer to floating-point conversions in pull request #1659 that haven't been integrated into the Google Highway trunk.

**Comment by fbarbari** [comment]
Ok, I see that you used a `hwy::SignedFromSize(sizeof(size_t))` tag to use a `ConvertTo`. I didn't know the existence of such tag. Anyway, thank you so much for your help and for the PR, I was going to ask why these functions were not included in the library but you somehow read in my mind :)
I am closing this issue, if it's ok for you.

**Comment by jan-wassenberg** [comment]
Thank you @johnplatts for adding them :) I've now reviewed the PR.

> I was going to ask why these functions were not included in the library 

We had started with the approach of only providing what is efficient on all platforms. However, sometimes you really do want that operation, and if the choice is between not vectorizing at all, or somewhat less efficient vector code, the latter usually wins.
Also, SVE/RVV/AVX-512 are quite similar in terms of features, and they all provide good support for these conversions, so it's great to add them now.
